mtp_bank_admin/apps/bank_admin/statement.py

```diff
@@ -0,0 +1,117 @@
+from bai2 import bai2, models, constants
+from moj_auth import api_client
+
+from . import BAI2_STMT_LABEL
+from .exceptions import EmptyFileError
+from .utils import retrieve_all_transactions, get_transaction_uid,\
+    create_batch_record
+
+SENDER_ID = 'NWBKGB2L'
+RECEIVER_ID = '391796'
+ACCOUNT_NUMBER = '10002383 607080'
+CURRENCY = 'GBP'
+RECORD_LENGTH = 80
+
+
+def generate_bank_statement(request):
+    transactions = retrieve_all_transactions(request, 'credited,refunded',
+                                             exclude_batch_label=BAI2_STMT_LABEL)
+
+    if len(transactions) == 0:
+        raise EmptyFileError()
+
+    transaction_records = []
+    credit_num = 0
+    credit_total = 0
+    debit_num = 0
+    debit_total = 0
+    for transaction in transactions:
+        transaction_record = models.TransactionDetail([])
+        if transaction.get('refunded', False):
+            transaction_record.type_code = constants.TypeCodes['699']
+            debit_num += 1
+            debit_total += transaction['amount']
+        else:
+            transaction_record.type_code = constants.TypeCodes['399']
+            credit_num += 1
+            credit_total += transaction['amount']
+
+        transaction_record.amount = transaction['amount']
+        transaction_record.text = str(get_transaction_uid(transaction))
+        transaction_records.append(transaction_record)
+
+    bai2_file = models.Bai2File()
+
+    file_header = bai2_file.header
+    file_header.sender_id = SENDER_ID
+    file_header.receiver_id = RECEIVER_ID
+    file_header.creation_date = datetime.date.today()
+    file_header.creation_time = datetime.datetime.utcnow().time()
+    file_header.file_id = int(time.time()) % 86400
+    file_header.physical_record_size = RECORD_LENGTH
+
+    group = models.Group()
+    group_header = group.header
+    group_header.ultimate_receiver_id = RECEIVER_ID
+    group_header.originator_id = SENDER_ID
+    group_header.group_status = constants.GroupStatus.update
+    group_header.as_of_date = datetime.date.today()
+    group_header.as_of_time = datetime.datetime.utcnow().time()
+    group_header.currency = CURRENCY
+    group_header.as_of_date_modifier = constants.AsOfDateModifier.interim_same_day
+    bai2_file.children.append(group)
+
+    # calculate balance values with reference to last created statement
+    opening_balance = 0
+    conn = api_client.get_connection(request)
+    response = conn.batches.get(label=BAI2_STMT_LABEL, limit=1)
+    if len(response['results']) > 0:
+        opening_balance = int(response['results'][0]['closing_balance'])
+    closing_balance = opening_balance + credit_total - debit_total
+
+    account = models.Account()
+    account_header = account.header
+    account_header.customer_account_number = ACCOUNT_NUMBER
+    account_header.currency = CURRENCY
+    account_header.summary_items = [
+        models.Summary(
+            type_code=constants.TypeCodes['010'],
+            amount=opening_balance
```

Comment:
While these `type_code` values are all clear from the context, it might be nice if we didn't have to have these magic strings (`010` and friends) here?